File: Java/String/SuffixArray.java
class SuffixArray
{
	char T[];                        
	int n, m;           
	int[] RA, tempRA; 
	Integer[] SuffArr, tempSA; 
	int[] Phi;  
	int[] PLCP;
	int[] LCP;    // LCP[i] stores the LCP between previous suffix "T + SuffArr[i-1]" and current suffix "T + SuffArr[i]"

	private void countingSort(int k)
	{
		int i, sum, maxi = Math.max(300, n);   // up to 255 ASCII chars or length of n
		int[] c = new int[maxi];
		for (i = 0; i < n; i++)
			c[i + k < n ? RA[i + k] : 0]++;
		for (i = sum = 0; i < maxi; i++) {
			int t = c[i]; c[i] = sum; sum += t;
		}
		for (i = 0; i < n; i++)               
			tempSA[c[SuffArr[i] + k < n ? RA[SuffArr[i] + k] : 0]++] = SuffArr[i];
		for (i = 0; i < n; i++)                    
			SuffArr[i] = tempSA[i];
	}

	private void constructSA() 
	{              
		int i, k, r;
		for (i = 0; i < n; i++) RA[i] = T[i];          //initial rank           
		for (i = 0; i < n; i++) SuffArr[i] = i;        
		for (k = 1; k < n; k <<= 1) {            
			countingSort(k);       
			countingSort(0);              
			tempRA[SuffArr[0]] = r = 0;                 
			for (i = 1; i < n; i++)                      
				tempRA[SuffArr[i]] = (RA[SuffArr[i]] == RA[SuffArr[i-1]] && RA[SuffArr[i]+k] == RA[SuffArr[i-1]+k]) ? r : ++r;
			for (i = 0; i < n; i++)                          // update the rank array RA
				RA[i] = tempRA[i];
		}		
	}

	private void computeLCP() 
	{
		int i, L;
		Phi[SuffArr[0]] = -1;
		for (i = 1; i < n; i++)
			Phi[SuffArr[i]] = SuffArr[i-1];         // remember which suffix is behind this suffix
		for (i = L = 0; i < n; i++) {                  // compute Permuted LCP in O(n)
			if (Phi[i] == -1) { PLCP[i] = 0; continue; }                 // special case
			while (i + L < T.length && Phi[i] + L < T.length && T[i + L] == T[Phi[i] + L]) L++;
			PLCP[i] = L;
			L = Math.max(L-1, 0);
		}
		for (i = 1; i < n; i++) 
			LCP[i] = PLCP[SuffArr[i]];   // put the permuted LCP back to the correct position
	}

	private int strncmp(char[] a, int i, char[] b, int j, int n){
		for (int k=0; i+k < a.length && j+k < b.length; k++){
			if (a[i+k] != b[j+k]) return a[i+k] - b[j+k];
		}
		return 0;
	}
	

	public int[] stringMatching(String str) // string matching in O(m log n)
	{
		char P[] = str.toCharArray();
		m = P.length;
		int lo = 0, hi = n-1, mid = lo;                 
		while (lo < hi) {                                          // find lower bound
			mid = (lo + hi) / 2;                                   
			int res = strncmp(T, SuffArr[mid], P, 0, m);
			if (res >= 0) hi = mid;            
			else          lo = mid + 1; 
		}
		if (strncmp(T,SuffArr[lo], P,0, m) != 0) return new int[]{-1, -1};         // if not found
		int[] ans = new int[]{ lo, 0} ;

		lo = 0; hi = n - 1; mid = lo;
		while (lo < hi) {                 // if lower bound is found, find upper bound
			mid = (lo + hi) / 2;
			int res = strncmp(T, SuffArr[mid], P,0, m);
			if (res > 0) hi = mid;
			else         lo = mid + 1;
		}
		if (strncmp(T, SuffArr[hi], P,0, m) != 0) hi--;                      // special case
		ans[1] = hi;
		return ans;// return lower/upper bound of the range where the pattern is found
	} 

	public String LRS() //longest repeated substring O(n)
	{
		int i, idx = 0, maxLCP = 0;

		for (i = 1; i < n; i++)
		if (LCP[i] > maxLCP) {
			maxLCP = LCP[i];
			idx = i;
		}
		return (new String(T)).substring(SuffArr[idx], SuffArr[idx]+maxLCP); // maxLCP is the length
	}

	private int owner(int idx) { return (idx < n-m-1) ? 1 : 2; }

	public String LCS(String str) 
	{
		int i, j, maxLCP = 0, idx = 0;
		m = str.length();
		T = (new String(T) + str + "#").toCharArray();   // append P and '#'
		n = T.length; // update n
		RA = new int[n];
		tempRA = new int[n];
		SuffArr = new Integer[n];
		tempSA = new Integer[n];
		Phi = new int[n];
		PLCP = new int[n];
		LCP = new int[n];
		constructSA();
		computeLCP();

		for (i = 1, maxLCP = -1; i < n; i++){
			if (LCP[i] > maxLCP && owner(SuffArr[i]) != owner(SuffArr[i-1])) {  // different owner
				maxLCP = LCP[i];
				idx = i;
			}
		}
		return (new String(T)).substring(SuffArr[idx], SuffArr[idx] + maxLCP);
	}
	
	public SuffixArray(String str)
	{
		T = str.toCharArray();
		n = T.length;
		RA = new int[n];
		tempRA = new int[n];
		SuffArr = new Integer[n];
		tempSA = new Integer[n];
		Phi = new int[n];
		PLCP = new int[n];
		LCP = new int[n];
		constructSA();
		computeLCP(); 
	}
}

File: C++/String/SuffixArray.cpp
struct SuffixArray{
	string s;                        
	int n, m;           
	vector<int> ra, tra, phi, plcp, lcp; 
	vector<int> sarr, tsa; 
	// LCP[i] stores the LCP between previous suffix "s + SuffArr[i-1]" and current suffix "s + SuffArr[i]"

	void csort(int k){
		int i, sum, maxi = max(300, n);   // up to 255 ASCII chars or length of n
		vector<int> c(maxi,0);
		for (i = 0; i < n; i++)
			c[i + k < n ? ra[i + k] : 0]++;
		for (i = sum = 0; i < maxi; i++){
			int t = c[i]; c[i] = sum; sum += t;
		}
		for (i = 0; i < n; i++)               
			tsa[c[sarr[i] + k < n ? ra[sarr[i] + k] : 0]++] = sarr[i];
		for (i = 0; i < n; i++)                    
			sarr[i] = tsa[i];
	}
	void constructSA() {              
		int i, k, r;
		for (i = 0; i < n; i++) ra[i] = s[i]; //initial rank           
		for (i = 0; i < n; i++) sarr[i] = i;        
		for (k = 1; k < n; k <<= 1) {            
			csort(k);       
			csort(0);              
			tra[sarr[0]] = r = 0;                 
			for (i = 1; i < n; i++)                      
				tra[sarr[i]] = (ra[sarr[i]] == ra[sarr[i-1]] && ra[sarr[i]+k] == ra[sarr[i-1]+k]) ? r : ++r;
			for (i = 0; i < n; i++) // update the rank array RA
				ra[i] = tra[i];
		}		
	}
	void computeLCP() {
		int i, l;
		phi[sarr[0]] = -1;
		for (i = 1; i < n; i++)
			phi[sarr[i]] = sarr[i-1];         // remember which suffix is behind this suffix
		for (i = l = 0; i < n; i++) {                  // compute Permuted LCP in O(n)
			if (phi[i] == -1) { plcp[i] = 0; continue; }                 // special case
			while (i + l < (int)s.size() && phi[i] + l < (int)s.size() && s[i + l] == s[phi[i] + l]) l++;
			plcp[i] = l;
			l = max(l-1, 0);
		}
		for (i = 1; i < n; i++) 
			lcp[i] = plcp[sarr[i]];   // put the permuted LCP back to the correct position
	}
	int strncmp(string a, int i, string b, int j, int n){
		for (int k=0; i+k < (int)a.size() && j+k < (int)b.size(); k++)
			if (a[i+k] != b[j+k]) return a[i+k] - b[j+k];
		return 0;
	}
	vector<int> stringMatching(string str){ // string matching in O(m log n)
		//char P[] = str.toCharArray();
		m = str.size();
		int lo = 0, hi = n-1, mid = lo;                 
		while (lo < hi) {                                          // find lower bound
			mid = (lo + hi) / 2;                                   
			int res = strncmp(s, sarr[mid], str, 0, m);
			if (res >= 0) hi = mid;            
			else          lo = mid + 1; 
		}
		if (strncmp(s,sarr[lo], str,0, m) != 0) return {-1, -1};         // if not found
		vector<int> ans = { lo, 0} ;
		lo = 0; hi = n - 1; mid = lo;
		while (lo < hi) {                 // if lower bound is found, find upper bound
			mid = (lo + hi) / 2;
			int res = strncmp(s, sarr[mid], str,0, m);
			if (res > 0) hi = mid;
			else         lo = mid + 1;
		}
		if (strncmp(s, sarr[hi], str,0, m) != 0) hi--;// special case
		ans[1] = hi;
		return ans;// return lower/upper bound of the range where the pattern is found
	} 
    string LongestRpeatedSubstring(){ //longest repeated substring O(n)
		int i, idx = 0, mam = 0;
		for (i = 1; i < n; i++)
		if (lcp[i] > mam) {
			mam = lcp[i];
			idx = i;
		}
		return s.substr(sarr[idx], mam); // maxLCP is the length
	}
	int owner(int idx) { return (idx < n-m-1) ? 1 : 2; }
	string LongestCommonSubstring(string str0, string str1) {
		int i, mam = 0, idx = 0;
		m = str1.length();
		s = str0 + str1 + "#";   // append P and '#'
		n = s.size(); // update n
		ra = vector<int>(n,0);
		tra = vector<int>(n,0);
		sarr = vector<int>(n,0);
		tsa = vector<int>(n,0);
		phi = vector<int>(n,0);
		plcp = vector<int>(n,0);
		lcp = vector<int>(n,0);
		constructSA();
		computeLCP();
		for (i = 1, mam = -1; i < n; i++){
			if (lcp[i] > mam && owner(sarr[i]) != owner(sarr[i-1])) {  // different owner
				mam = lcp[i];
				idx = i;
			}
		}
		return s.substr(sarr[idx], mam);
	}
	void suffixarray(string str){
		s = str+"$";
		n = s.size();
		ra = vector<int>(n,0);
		tra = vector<int>(n,0);
		sarr = vector<int>(n,0);
		tsa = vector<int>(n,0);
		phi = vector<int>(n,0);
		plcp = vector<int>(n,0);
		lcp = vector<int>(n,0);
		constructSA();
		computeLCP(); 
	}
};